src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp

```diff
@@ -756,6 +765,7 @@ void ShenandoahControlThread::handle_requested_gc(GCCause::Cause cause) {
   while (current_gc_id < required_gc_id) {
     _gc_requested.set();
     _requested_gc_cause = cause;
+    notify_control_thread();
```

Comment:
So this thing holds `_gc_waiters_lock`. I noticed it when trying to implement JDK-8264851. Fastdebug builds would fail with: `#  assert(false) failed: Attempting to acquire lock ShenandoahControlWaitGC_lock/11 out of order with lock ShenandoahRequestedGC_lock/11 -- possible deadlock`. How is that handled here?

Reply:
Looking into this.

Reply:
I think this is a false positive. There's no deadlock here because the control thread never tries to hold both these locks at the same time. I changed the rank of the `_gc_waiters_lock` to allow a thread to hold them both.